### PR TITLE
[BEAM-6756] Create Iterable type for Schema

### PR DESCRIPTION
**Pull request by reuvenlax**
Unlike the existing Array type, an Iterable type does not have to be materialized into memory. This allows the result of GroupByKey and CoGroupByKey to be modeled with schemas.

### CODE REVIEW

**Comment by reuvenlax** [comment]
run dataflow validatesrunner

**Comment by reuvenlax** [comment]
run sql postcommit

**Comment by reuvenlax** [comment]
R: @TheNeuralBit

File: learning/katas/java/.idea/study_project.xml [hunk 1]
```diff
@@ -0,0 +1,3151 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (TheNeuralBit) [inline learning/katas/java/.idea/study_project.xml]: Is this related?

Comment (reuvenlax) [reply learning/katas/java/.idea/study_project.xml]: No. Ignore that file - I will remove it from the PR.

**Comment by reuvenlax** [comment]
run sql postcommit

**Comment by reuvenlax** [comment]
friendly ping

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java [hunk 2]
```diff
@@ -292,7 +293,7 @@ private static Schema fromTableFieldSchema(List<TableFieldSchema> tableFieldSche
       if (!schemaField.getType().getNullable()) {
         field.setMode(Mode.REQUIRED.toString());
       }
-      if (TypeName.ARRAY == type.getTypeName()) {
+      if (type.getTypeName().isCollectionType()) {
```
Comment (TheNeuralBit) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java]: Is there a chance there are other instances of checks like this that need to be replaced?

**Comment by reuvenlax** [comment]
run dataflow validatesrunner

**Comment by reuvenlax** [comment]
run sql postcommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java (hunk 2 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java]: I did another deep search - I found one case in SQL's codegen which I fixed, but can't find anything else.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java [hunk 3]
```diff
@@ -450,28 +469,28 @@ static int deepHashCodeForMap(
...
-    static boolean deepEqualsForList(List<Object> a, List<Object> b, Schema.FieldType elementType) {
+    static boolean deepEqualsForIterable(
+        Iterable<Object> a, Iterable<Object> b, Schema.FieldType elementType) {
       if (a == b) {
         return true;
       }
 
-      if (a.size() != b.size()) {
-        return false;
-      }
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java]: good catch - fixed.

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit
